File: anika.dovob/src/main/java/com/example/anika/homework/chapter7/OrderItem.java
package com.example.anika.homework.chapter7;

import jakarta.persistence.*;

import java.util.HashSet;
import java.util.Set;

@Entity(name = "Ch7OrderItem")
@Table(name = "order_item_ch7")
public class OrderItem extends BaseEntity {

    @Id
    @GeneratedValue
    @Column(name = "order_item_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id", nullable = false)
    private Order order;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_id")
    private Item item;

    @Column(name = "item_name_snapshot", nullable = false)
    private String itemNameSnapshot;

    @Column(name = "unit_price_snapshot", nullable = false)
    private long unitPriceSnapshot;

    @Column(nullable = false)
    private int quantity;

    @ElementCollection
    @CollectionTable(
            name = "order_item_option_ch7",
            joinColumns = @JoinColumn(name = "order_item_id")
    )
    private Set<ItemOption> options = new HashSet<>();

    protected OrderItem() {
    }

    public OrderItem(Item item, int quantity) {
        this.item = item;
        this.itemNameSnapshot = item.getName();
        this.unitPriceSnapshot = item.getBasePrice();
        this.quantity = quantity;
    }

    public void addOption(ItemOption option) {
        options.add(option);
    }

    public Long getId() {
        return id;
    }

    public Order getOrder() {
        return order;
    }

    void setOrder(Order order) {
        this.order = order;
    }

    public Item getItem() {
        return item;
    }

    public String getItemNameSnapshot() {
        return itemNameSnapshot;
    }

    public long getUnitPriceSnapshot() {
        return unitPriceSnapshot;
    }

    public int getQuantity() {
        return quantity;
    }

    public Set<ItemOption> getOptions() {
        return options;
    }
}

File: hyun.k/hyun.k/src/main/kotlin/chapter7/OrderItem.kt
package chapter7

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany

// 과제 내용을 이해하기론, Order와 Item의 다대다 매핑을 중간 엔티티를 통해 풀어내는 부분을 의도하신게 아닌가..
@Entity
class OrderItem(
) : BaseEntity() {
    // TODO: 식별 or 비식별 선택
    // ㄴ 비식별로 선택 (자체 pk 가짐)

    @Id
    @GeneratedValue
    @Column(name = "order_item_id")
    var id: Long? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id")
    var order: Order? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "item_id")
    var item: Item? = null

    var itemNameSnapshot: String = ""
    var unitPriceSnapshot: Long = 0
    var quantity: Int = 0

    @OneToMany(mappedBy = "orderItem")
    var options = mutableListOf<OrderItemOption>()
}